src/Server/ArrowFlightHandler.cpp

```diff
@@ -0,0 +1,526 @@
+#include <Parsers/ASTIdentifier.h>
+#include <Access/Credentials.h>
+#include <Processors/Formats/Impl/CHColumnToArrowColumn.h>
+#include <Common/Base64.h>
+#include <arrow/flight/server_middleware.h>
+#include <Poco/Util/LayeredConfiguration.h>
+#include <arrow/flight/server.h>
+#include <fstream>
+#include <sstream>
+#include <Poco/FileStream.h>
+#include <Poco/StreamCopier.h>
+
+const std::string AUTHORIZATION_HEADER = "authorization";
+const std::string AUTHORIZATION_MIDDLEWARE_NAME = "arrow_flight_authorization";
+
+class AuthMiddleware : public arrow::flight::ServerMiddleware {
+public:
+    explicit AuthMiddleware(const std::string& token, const std::string& username, const std::string& password):
+        token_(token),
+        username_(username),
+        password_(password) {}
+        
+    const std::string& username() const { return username_; }
+    const std::string& password() const { return password_; }
+
+    std::string name() const override {
+        return AUTHORIZATION_MIDDLEWARE_NAME;
+    }
+
+    void SendingHeaders(arrow::flight::AddCallHeaders* outgoing_headers) override {
+        outgoing_headers->AddHeader(AUTHORIZATION_HEADER, "Bearer " + token_);
+    }
+
+    void CallCompleted(const arrow::Status& /*status*/) override {
+    }
+
+private:
+    const std::string token_;
+    const std::string username_;
+    const std::string password_;
+};
+
+class AuthMiddlewareFactory : public arrow::flight::ServerMiddlewareFactory {
+public:
+    arrow::Status StartCall(const arrow::flight::CallInfo& /*info*/, const arrow::flight::ServerCallContext& context,
+                           std::shared_ptr<arrow::flight::ServerMiddleware>* middleware) override {
+        const auto& headers = context.incoming_headers();
+
+        auto it = headers.find(AUTHORIZATION_HEADER);
+        if (it == headers.end()) {
+            return arrow::Status::IOError("Missing Authorization header");
+        }
+
+        auto auth_header = std::string(it->second);
+
+        std::string token;
+
+        const std::string prefix_basic = "Basic ";
+        if (auth_header.compare(0, prefix_basic.size(), prefix_basic) == 0) {
+            token = auth_header.substr(prefix_basic.size());
+        }
+
+        const std::string prefix_bearer = "Bearer ";
+        if (auth_header.compare(0, prefix_bearer.size(), prefix_bearer) == 0) {
+            token = auth_header.substr(prefix_bearer.size());
+        } 
+
+        if (token.empty()) {
+            return arrow::Status::IOError("Expected Basic auth scheme");
+        }
+
+        std::string credentials = DB::base64Decode(token, true);
+        auto pos = credentials.find(':');
+        if (pos == std::string::npos) {
+            return arrow::Status::IOError("Malformed credentials");
+        }
+
+        auto user = credentials.substr(0, pos);
+        auto password = credentials.substr(pos + 1);
+
+        std::cout << token << " : " << user << " : " << password << "\n";
```

Comment:
Please remove this - it's bad if someone finds your password in the logs.

Reply:
fixed